Zend/zend_attributes.h

```diff
@@ -0,0 +1,82 @@
+#ifndef ZEND_ATTRIBUTES_H
+#define ZEND_ATTRIBUTES_H
+
+#define ZEND_ATTRIBUTE_TARGET_CLASS			1
+#define ZEND_ATTRIBUTE_TARGET_FUNCTION		2
+#define ZEND_ATTRIBUTE_TARGET_METHOD		4
+#define ZEND_ATTRIBUTE_TARGET_PROPERTY		8
+#define ZEND_ATTRIBUTE_TARGET_CLASS_CONST	16
+#define ZEND_ATTRIBUTE_TARGET_PARAMETER		32
+#define ZEND_ATTRIBUTE_TARGET_ALL			63
+
+#define ZEND_ATTRIBUTE_SIZE(argc) (sizeof(zend_attribute) + sizeof(zval) * (argc) - sizeof(zval))
+
+BEGIN_EXTERN_C()
+
+extern ZEND_API zend_class_entry *zend_ce_php_attribute;
+extern ZEND_API zend_class_entry *zend_ce_php_compiler_attribute;
+
+typedef struct _zend_attribute {
+	zend_string *name;
+	zend_string *lcname;
+	uint32_t offset;
+	uint32_t argc;
+	zval argv[1];
+} zend_attribute;
+
+typedef void (*zend_attributes_internal_validator)(zend_attribute *attr, int target);
+
+static zend_always_inline void zend_attribute_free(zend_attribute *attr)
+{
+	uint32_t i;
+
+	zend_string_release(attr->name);
+	zend_string_release(attr->lcname);
+
+	for (i = 0; i < attr->argc; i++) {
+		zval_ptr_dtor(&attr->argv[i]);
+	}
+
+	efree(attr);
+}
+
+static zend_always_inline zend_attribute *zend_get_attribute(HashTable *attributes, zend_string *name, uint32_t offset)
```

Comment:
Similarly here, I don't think these needs to be zend_always_inline. This code does not seem particularly performance critical to me.

Reply:
```suggestion
static zend_always_inline zend_attribute *zend_get_attribute(HashTable *attributes, zend_string *lcname, uint32_t offset)
```

To make it clear that it must be lowercased beforehand.